File: packages/eixam_connect_flutter/android/src/main/kotlin/dev/eixam/connect/flutter/protection/TelAggregateReassembler.kt
package dev.eixam.connect.flutter.protection

/**
 * Mirrors Dart [EixamTelFragment] / [EixamTelReassembler]. Nearby text is
 * 22 B header + UTF-8, so it always arrives as 0xD0 15-byte chunks on TEL.
 */
internal data class TelAggregateFragment(
    val totalLength: Int,
    val offset: Int,
    val fragmentPayload: List<Int>,
) {
    val fragmentLength: Int get() = fragmentPayload.size

    companion object {
        private const val opcode = 0xD0
        private const val headerLength = 5
        private const val maxPayloadLength = 15

        fun tryParse(bytes: List<Int>): TelAggregateFragment? {
            if (bytes.size < headerLength + 1 || bytes.first() != opcode) {
                return null
            }
            val totalLength = (bytes[1] and 0xFF) or ((bytes[2] and 0xFF) shl 8)
            val offset = (bytes[3] and 0xFF) or ((bytes[4] and 0xFF) shl 8)
            val payload = bytes.subList(headerLength, bytes.size)
            if (totalLength <= 0 || payload.isEmpty() || payload.size > maxPayloadLength) {
                return null
            }
            return TelAggregateFragment(totalLength, offset, payload.toList())
        }
    }
}

internal class TelAggregateReassembler {
    private var activeTotalLength: Int? = null
    private val fragmentsByOffset = linkedMapOf<Int, List<Int>>()

    fun ingest(payload: List<Int>): List<Int>? {
        val fragment = TelAggregateFragment.tryParse(payload) ?: return payload
        return addFragment(fragment)
    }

    fun addFragment(fragment: TelAggregateFragment): List<Int>? {
        if (fragment.offset < 0) {
            reset()
            return null
        }
        val active = activeTotalLength
        if (active == null) {
            activeTotalLength = fragment.totalLength
        } else if (active != fragment.totalLength) {
            reset()
            activeTotalLength = fragment.totalLength
        }
        val fragmentEnd = fragment.offset + fragment.fragmentLength
        if (fragmentEnd > fragment.totalLength) {
            reset()
            return null
        }
        for ((existingStart, existingPayload) in fragmentsByOffset) {
            val existingEnd = existingStart + existingPayload.size
            val overlaps = fragment.offset < existingEnd && fragmentEnd > existingStart
            if (!overlaps) {
                continue
            }
            val sameRange = existingStart == fragment.offset &&
                existingEnd == fragmentEnd &&
                existingPayload == fragment.fragmentPayload
            if (sameRange) {
                return tryComplete(fragment.totalLength)
            }
            reset()
            return null
        }
        fragmentsByOffset[fragment.offset] = fragment.fragmentPayload
        return tryComplete(fragment.totalLength)
    }

    fun reset() {
        activeTotalLength = null
        fragmentsByOffset.clear()
    }

    private fun tryComplete(totalLength: Int): List<Int>? {
        if (fragmentsByOffset.isEmpty()) {
            return null
        }
        val ordered = fragmentsByOffset.keys.sorted()
        var cursor = 0
        val completed = ArrayList<Int>(totalLength)
        for (offset in ordered) {
            if (offset != cursor) {
                return null
            }
            val payload = fragmentsByOffset[offset] ?: return null
            completed.addAll(payload)
            cursor += payload.size
        }
        if (cursor != totalLength) {
            return null
        }
        reset()
        return completed
    }
}

File: packages/eixam_connect_flutter/ios/Classes/TelAggregateReassembler.swift
import Foundation

struct TelAggregateFragment {
  let totalLength: Int
  let offset: Int
  let fragmentPayload: [UInt8]

  var fragmentLength: Int { fragmentPayload.count }

  static func tryParse(_ bytes: [UInt8]) -> TelAggregateFragment? {
    guard bytes.count >= 6, bytes[0] == 0xD0 else {
      return nil
    }
    let totalLength = Int(bytes[1]) | (Int(bytes[2]) << 8)
    let offset = Int(bytes[3]) | (Int(bytes[4]) << 8)
    let payload = Array(bytes[5...])
    guard totalLength > 0, !payload.isEmpty, payload.count <= 15 else {
      return nil
    }
    return TelAggregateFragment(
      totalLength: totalLength,
      offset: offset,
      fragmentPayload: payload
    )
  }
}

final class TelAggregateReassembler {
  private var activeTotalLength: Int?
  private var fragmentsByOffset: [Int: [UInt8]] = [:]

  func ingest(_ payload: [UInt8]) -> [UInt8]? {
    guard let fragment = TelAggregateFragment.tryParse(payload) else {
      return payload
    }
    return addFragment(fragment)
  }

  func addFragment(_ fragment: TelAggregateFragment) -> [UInt8]? {
    if fragment.offset < 0 {
      reset()
      return nil
    }
    if let active = activeTotalLength {
      if active != fragment.totalLength {
        reset()
        activeTotalLength = fragment.totalLength
      }
    } else {
      activeTotalLength = fragment.totalLength
    }
    let fragmentEnd = fragment.offset + fragment.fragmentLength
    if fragmentEnd > fragment.totalLength {
      reset()
      return nil
    }
    for (existingStart, existingPayload) in fragmentsByOffset {
      let existingEnd = existingStart + existingPayload.count
      let overlaps = fragment.offset < existingEnd && fragmentEnd > existingStart
      guard overlaps else {
        continue
      }
      let sameRange = existingStart == fragment.offset &&
        existingEnd == fragmentEnd &&
        existingPayload == fragment.fragmentPayload
      if sameRange {
        return tryComplete(fragment.totalLength)
      }
      reset()
      return nil
    }
    fragmentsByOffset[fragment.offset] = fragment.fragmentPayload
    return tryComplete(fragment.totalLength)
  }

  func reset() {
    activeTotalLength = nil
    fragmentsByOffset.removeAll()
  }

  private func tryComplete(_ totalLength: Int) -> [UInt8]? {
    if fragmentsByOffset.isEmpty {
      return nil
    }
    let ordered = fragmentsByOffset.keys.sorted()
    var cursor = 0
    var completed: [UInt8] = []
    completed.reserveCapacity(totalLength)
    for offset in ordered {
      if offset != cursor {
        return nil
      }
      guard let payload = fragmentsByOffset[offset] else {
        return nil
      }
      completed.append(contentsOf: payload)
      cursor += payload.count
    }
    if cursor != totalLength {
      return nil
    }
    reset()
    return completed
  }
}
